py-rattler/src/networking/client.rs

```diff
@@ -0,0 +1,39 @@
+use crate::networking::middleware::PyMiddleware;
+use pyo3::{pyclass, pymethods};
+use rattler_networking::{AuthenticationMiddleware, MirrorMiddleware};
+use reqwest_middleware::ClientWithMiddleware;
+
+#[pyclass]
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct PyClientWithMiddleware {
+    pub(crate) inner: ClientWithMiddleware,
+}
+
+#[pymethods]
+impl PyClientWithMiddleware {
+    #[new]
+    pub fn new(middlewares: Option<Vec<PyMiddleware>>) -> Self {
+        let middlewares = middlewares.unwrap_or_default();
+        let client = reqwest_middleware::ClientBuilder::new(reqwest::Client::new());
+        let client = middlewares
+            .into_iter()
+            .fold(client, |client, middleware| match middleware {
+                PyMiddleware::MirrorMiddleware(middleware) => {
+                    client.with(MirrorMiddleware::from(middleware))
```

Comment:
Can we also use `middleware.into()` here and below?

Reply:
nope

```
error[E0283]: type annotations needed
  --> src/networking/client.rs:23:28
   |
23 |                     client.with(middleware.into())
   |                            ^^^^ ----------------- type must be known at this point
   |                            |
   |                            cannot infer type of the type parameter `M` declared on the method `with`
```

also i find `MirrorMiddleware::from` a bit easier to understand in this scenario

Reply:
although `client.with::<MirrorMiddleware>(middleware.into())` works

Reply:
but i still find `::from` a bit nicer to read so i'll leave it as is since we are not able to collapse it all into one line using `into` anyway